### ISSUE TITLE
Support inline "input method" input

### ISSUE THREAD

**Issue by crocket** [jwilm/alacritty#1613]
Which operating system does the issue occur on? Gentoo Linux
If on linux, are you using X11 or Wayland? X11

On alacritty

![2018-10-01_mon_23 53 52](https://user-images.githubusercontent.com/748856/46321951-4fcf3480-c621-11e8-8010-4f5cad40b792.png)

On termite

![2018-10-01_mon_23 58 54](https://user-images.githubusercontent.com/748856/46321968-6d9c9980-c621-11e8-87b5-f77598e15cb4.png)

Out-of-band input method input is disorienting.

**Comment by crocket** [comment]
Is there a way to solve this issue without incorporating gtk or qt?

**Comment by chrisduerr** [comment]
It is possible to solve this without gtk or qt, there's no reason why this would be necessary.

However making these changes inline is complicated, since Alacritty doesn't handle this directly and there's no logic to support it at all.

**Comment by pickfire** [comment]
Alacritty seemed to support over-the-spot IME, is there any reason why over-the-spot preedit area is not good enough to cover usage like on-the-spot preedit area?

![2020-05-19-225603_1366x768_scrot](https://user-images.githubusercontent.com/4687791/82342421-2b0d1800-9a24-11ea-9dfa-eed2ad49d1c9.png)

It seemed to be common to have over-the-spot preedit area in X11, I never used on-the-spot preedit area. Maybe it performs better for chinese input rather than japanese input.

**Comment by AitBits** [comment]
This would be a nice QoL addition to CJK users, because over-the-spot preedit often feels rather clunky, since you have to consciously switch focus up and down.

**Comment by pickfire** [comment]
@AitBits No, I never had to do that. I never looked at the place where we edit, I just look at the preedit area, maybe japanese is different since one seemed to be able to insert a few characters at a time, for chinese most of time I insert small words since it cannot insert character one by one.

**Comment by keichi** [comment]
On-the-spot edit has been the default on Japanese environments too. Alacritty seems really nice but I am unable to switch from iTerm2 because it lacks inline edit.

Alacritty
![alacritty](https://user-images.githubusercontent.com/791283/87317660-4a2ab080-c562-11ea-8bca-bb98b96183ae.png)

iTerm2
![iterm2](https://user-images.githubusercontent.com/791283/87317669-4dbe3780-c562-11ea-951c-9bc9cdcf1ff1.png)

**Comment by crocket** [comment]
Inline input method is ideal, but I'd be happy with wayland input method support.

Right now, alacritty doesn't support wayland input method standards.

**Comment by kchibisov** [comment]
alacritty has support for `text-input-v3` and last time it worked on gnome just fine.

**Comment by crocket** [comment]
Fcitx5 doesn't work on alacritty with wayland backend.

**Comment by kchibisov** [comment]
Does fcitx5 setup to use Wayland backend and not a dbus? I know that it should work, but last time sway/fcitx has popup window problems.

**Comment by crocket** [comment]
I don't know whether fcitx5 is configured to use wayland backend.

**Comment by crocket** [comment]
fcitx5 kind of works on alacritty after disabling a lot of frontends in fcitx5.

But, disabling the frontends breaks qt applications and gtk applications.

Also, wayland frontend doesn't display pre-edit text in application and lets me activate only the first input method. I have multiple input methods.

**Comment by crocket** [comment]
After further testing, I confirmed that disabling classic user interface in fcitx5 enables input method on wayland alacritty.

However, if I disable classic user interface, input method is extremely buggy on alacritty.
It is too buggy to be useable.

**Comment by kchibisov** [comment]
If someone could test inline input method implemented in #5790 that will be nice, though only X11 and Wayland are supported.

**Comment by KeenS** [comment]
Thank you @kchibisov for tackling this problem. I tried it on Ubuntu 21.10 on Xorg with uim-skk. While inline candidates displayed in inline, box candidates are displayed below the window.
![image](https://user-images.githubusercontent.com/4434568/149878991-72e5b999-cb04-4b61-a04f-a3e507588d1a.png)
![2022-01-18 14 57 13](https://user-images.githubusercontent.com/4434568/149879357-4e56077b-f6bf-47d4-8dac-96094cb6ccae.jpg)

**Comment by kchibisov** [comment]
@KeenS does master has the same issue? I can't test X11 now, but I can say that on Wayland they aligned properly.

**Comment by KeenS** [comment]
Yes. Block candidate displayed outside of window is almost the same behavior as master's.
![image](https://user-images.githubusercontent.com/4434568/149883001-602fc501-b944-4441-a41f-bb998dfb363f.png)
![2022-01-18 15 31 31](https://user-images.githubusercontent.com/4434568/149883125-c106dbe8-de6a-4942-a637-4d152bdd40e3.jpg)

BTW, I found that input is duplicated when the conversion is submitted. It doesn't occur in master

![image](https://user-images.githubusercontent.com/4434568/149883202-3b2821ca-e088-4dbb-a1b2-e35380447528.png)

**Comment by kchibisov** [comment]
@garasubo Would you mind looking on X11 bugs? Since I can't repro any of that on Wayland, and sending the same ime position results in different ime pop hotspot positions. (I've tested on gnome/x11).

![x11](https://user-images.githubusercontent.com/27620401/149884068-327183f3-951f-4820-9599-578c83f1191b.png)

**Comment by garasubo** [comment]
@kchibisov Sure. It looks the bug in winit level. It seems broken when rebasing the IME branch to the latest master, but couldn't find the root cause.

**Comment by kchibisov** [comment]
Ok, mentioned PR should work on X11, Wayland, and macOS. Tests are welcome.

Note, on X11 preedit popup window is wrong positioned due to limitations of libX11(patch was sent, but not sure it'll get further there).

**Comment by nullchilly** [comment]
Hope this lands in version 0.11, will definitely move to alacritty after this!

**Comment by crocket** [comment]
foot obtained inline wayland input method, but sway still needs to support candidate boxes for input methods.

**Comment by lolicon0930** [comment]
CJK users, the patch doesn't seem to work?
![2022-08-30-00:55:02](https://user-images.githubusercontent.com/14966910/187254087-e0f495a8-bbd8-41aa-8b2b-d62d6b37f31c.png)
```
$fcitx5 -v
5.0.18
```

**Comment by kchibisov** [comment]
Do you have inline preedit or you just have that popup at the bottom of the screen all the time. Could you provide the `--print-events` log if you don't have inline editing?

**Comment by lolicon0930** [comment]
I have set up inline preedit in program globally.
![2022-08-30-04:43:04](https://user-images.githubusercontent.com/14966910/187296395-66bb7cfa-3dce-4359-92fd-bec63746fab3.png)
If I cancel it.
![2022-08-30-04:43:59](https://user-images.githubusercontent.com/14966910/187296503-a8aa4a4c-67b2-4651-b2a5-e7a12a95e307.png)
![2022-08-30-04:44:13](https://user-images.githubusercontent.com/14966910/187296514-f239c3b4-1400-4149-a704-56067ac03817.png)
Here is the log when inline is set up, with `--print-events` : 
[Alacritty-55981.log](https://github.com/alacritty/alacritty/files/9447910/Alacritty-55981.log)

**Comment by kchibisov** [comment]
It seems like your IME provide its own inline input and given that we Xlib doesn't support popup positioning you have it that way. Could you setup your IME to operate normally with preedit so it won't try to draw preedit and basically forward events to alacritty? The issue is that your IME does its own inline preedit, so it's not really inline preedit in the application.

The behavior you're seeing is sort of expected on X11 given that Xlib doesn't support ime popup positioning when preedit is being used and we can't move away from Xlib for now...

**Comment by lolicon0930** [comment]
It works now, thanks all.
![2022-08-31-17:41:02-s](https://user-images.githubusercontent.com/14966910/187650265-1532c552-e521-4216-97e6-605685290b81.png)

**Comment by pickfire** [comment]
@lolicon0930 I guess it's a bug from fcitx5?

**Comment by kchibisov** [comment]
It was a combination of us doing weird, but valid thing and fcitx not thinking that it's possible and not throwing an error here. In general we'll do what fcitx author suggested, but it's sort of late to fix it for the next release, since it's in winit and would require breaking release already.
